File: lib/java/src/main/java/uapi/internal/schema/ParseStructFields.java
package uapi.internal.schema;

import static uapi.internal.schema.ParseField.parseField;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import uapi.UApiSchemaParseError;
import uapi.internal.types.UFieldDeclaration;

public class ParseStructFields {
    static Map<String, UFieldDeclaration> parseStructFields(
            List<Object> path,
            Map<String, Object> referenceStruct, ParseContext ctx) {
        final var parseFailures = new ArrayList<SchemaParseFailure>();
        final var fields = new HashMap<String, UFieldDeclaration>();

        for (final var structEntry : referenceStruct.entrySet()) {
            final var fieldDeclaration = structEntry.getKey();

            for (final var existingField : fields.keySet()) {
                final var existingFieldNoOpt = existingField.split("!")[0];
                final var fieldNoOpt = fieldDeclaration.split("!")[0];
                if (fieldNoOpt.equals(existingFieldNoOpt)) {
                    final List<Object> finalPath = new ArrayList<>(path);
                    finalPath.add(fieldDeclaration);

                    final List<Object> finalOtherPath = new ArrayList<>(path);
                    finalOtherPath.add(existingField);

                    final var finalOtherDocumentJson = ctx.uApiSchemaDocumentNamesToJson.get(ctx.documentName);
                    final var finalOtherLocationPseudoJson = GetPathDocumentCoordinatesPseudoJson
                            .getPathDocumentCoordinatesPseudoJson(finalOtherPath, finalOtherDocumentJson);

                    parseFailures
                            .add(new SchemaParseFailure(ctx.documentName, finalPath, "PathCollision",
                                    Map.of("document", ctx.documentName, "path", finalOtherPath, "location",
                                            finalOtherLocationPseudoJson)));
                }
            }

            final var typeDeclarationValue = structEntry.getValue();

            final UFieldDeclaration parsedField;
            try {
                parsedField = parseField(path, fieldDeclaration,
                        typeDeclarationValue, ctx);
                final String fieldName = parsedField.fieldName;

                fields.put(fieldName, parsedField);
            } catch (UApiSchemaParseError e) {
                parseFailures.addAll(e.schemaParseFailures);
            }
        }

        if (!parseFailures.isEmpty()) {
            throw new UApiSchemaParseError(parseFailures, ctx.uApiSchemaDocumentNamesToJson);
        }

        return fields;
    }
}

File: lib/py/uapi/internal/schema/ParseStructFields.py
from typing import TYPE_CHECKING

from uapi.internal.schema.SchemaParseFailure import SchemaParseFailure

if TYPE_CHECKING:
    from uapi.internal.schema.ParseContext import ParseContext
    from uapi.internal.types.UFieldDeclaration import UFieldDeclaration


def parse_struct_fields(reference_struct: dict[str, object], ctx: 'ParseContext') -> dict[str, 'UFieldDeclaration']:
    from uapi.UApiSchemaParseError import UApiSchemaParseError
    from uapi.internal.schema.ParseField import parse_field

    parse_failures = []
    fields: dict[str, 'UFieldDeclaration'] = {}

    for field_declaration, type_declaration_value in reference_struct.items():
        for existing_field in fields.keys():
            existing_field_no_opt = existing_field.split("!")[0]
            field_no_opt = field_declaration.split("!")[0]
            if field_no_opt == existing_field_no_opt:
                final_path = ctx.path + [field_declaration]
                final_other_path = ctx.path + [existing_field]
                parse_failures.append(SchemaParseFailure(
                    ctx.document_name, final_path, "PathCollision", {"document": ctx.document_name, "path": final_other_path}))

        try:
            parsed_field = parse_field(
                field_declaration, type_declaration_value, ctx)
            field_name = parsed_field.field_name
            fields[field_name] = parsed_field
        except UApiSchemaParseError as e:
            parse_failures.extend(e.schema_parse_failures)

    if parse_failures:
        raise UApiSchemaParseError(parse_failures)

    return fields
